consensus/reactor.go

```diff
@@ -858,31 +1005,67 @@ func (ps *PeerState) ApplyCommitStepMessage(msg *CommitStepMessage) {
 	ps.ProposalBlockParts = msg.BlockParts
 }
 
-func (ps *PeerState) ApplyHasVoteMessage(msg *HasVoteMessage) {
+func (ps *PeerState) ApplyProposalPOLMessage(msg *ProposalPOLMessage) {
 	ps.mtx.Lock()
 	defer ps.mtx.Unlock()
 
 	if ps.Height != msg.Height {
 		return
 	}
+	if ps.ProposalPOLRound != msg.ProposalPOLRound {
+		return
+	}
 
-	ps.setHasVote(msg.Height, msg.Round, msg.Type, msg.Index)
+	// TODO: Merge onto existing ps.ProposalPOL?
+	// We might have sent some prevotes in the meantime.
+	ps.ProposalPOL = msg.ProposalPOL
 }
 
-func (ps *PeerState) ApplyProposalPOLMessage(msg *ProposalPOLMessage) {
+func (ps *PeerState) ApplyHasVoteMessage(msg *HasVoteMessage) {
 	ps.mtx.Lock()
 	defer ps.mtx.Unlock()
 
 	if ps.Height != msg.Height {
 		return
 	}
-	if ps.ProposalPOLRound != msg.ProposalPOLRound {
-		return
+
+	ps.setHasVote(msg.Height, msg.Round, msg.Type, msg.Index)
+}
+
+// When a peer claims to have a maj23 for some BlockID at H,R,S,
+// we will try to respond with a VoteSetBitsMessage showing which
+// bits we already have (and which we don't yet have),
+// but that happens in another goroutine.
```

Comment:
Why bother with an extra go-routine per peer? Is there some reason we can't TrySend the VoteSetBits from Receive?